### PR TITLE
Add "Inspect Native Shader Code" to shader inspector and shader editor

### LINKED ISSUE #10278: Make "inspect native shader code" available also when selecting shader resources
### Describe the project you are working on

Godot engine

### Describe the problem or limitation you are having in your project

"Inspect native shader code" is available when inspecting a material, but not when inspecting a shader

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Make it avialable for shaders too

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Add the option in the dropdown

### If this enhancement will not be used often, can it be worked around with a few lines of script?

No

### Is there a reason why this should be core and not an add-on in the asset library?

Native editor feature

### PR DESCRIPTION
**Pull request by tetrapod00**
Implements and closes https://github.com/godotengine/godot-proposals/issues/10278.

Adds the action "Inspect Native Shader Code", currently available on the [Material resource](https://docs.godotengine.org/en/stable/classes/class_material.html#class-material-method-inspect-native-shader-code) inspector ([implementation](https://github.com/godotengine/godot/blob/0a4aedb36065f66fc7e99cb2e6de3e55242f9dfb/scene/resources/material.cpp#L106C1-L112C2)), to the Shader resource inspector and to the shader editor.
LMPrAaIzQi
SC05RjXnlO

godot windows editor dev x86_64_tU88Nvc2Ah
godot windows editor dev x86_64_hcpELW4Tsh
godot windows editor dev x86_64_OBXfqaaZUL

I added the action to the shader editor under the `File` menu, because it's a valid action for both text and visual shaders. I can see reasons to instead do one of the following:
- Add the action to one of the other menus, like `Edit`, which only appear for text shaders. This action is for advanced users, currently takes multiple seconds to open on my machine, and may be confused with the "Show Generated Shader Code" action in visual shaders. On the other hand, it is a *valid* action for both types of shader.
- Not include this action in the shader editor at all. Perhaps it will cause more confusion than its worth. With this PR, you can now do `File -> Open File in Inspector` and then `Manage object properties -> Inspect Native Shader Code`.

### CODE REVIEW

**Review by Calinou — APPROVED** [review]
Tested locally on both text and visual shaders, it works as expected.

File: scene/resources/shader.cpp
Comment (Calinou) [inline scene/resources/shader.cpp:261]: Is there a strong use case for exposing this to the scripting API? You should be able to use this within the editor without exposing it. It's only needed to expose it for editor plugins, but I'm not sure why they'd need to programmatically spawn the dialog.

Comment (tetrapod00) [reply scene/resources/shader.cpp:261]: I copied the existing implementation on Material. I assumed there was some reason it was done the way it is:
https://github.com/godotengine/godot/blob/0a4aedb36065f66fc7e99cb2e6de3e55242f9dfb/scene/resources/material.cpp#L150-L151

**Review by clayjohn — APPROVED** [review]
Seems fine to me!

File: scene/resources/shader.cpp
Comment (clayjohn) [reply scene/resources/shader.cpp:261]: That looked strange to me too, but I don't really understand how those flags work. I guess "DEFAULT" allows it to be called from scripting, but "EDITOR" limits that to an editor build?

At any rate, its fine since this is indeed copy-pasted from the existing implementation

Comment (tetrapod00) [reply scene/resources/shader.cpp:261]: And, to state the maybe not-so-obvious - there's no good way I saw for the `Material` implementation call the `Shader` implementation, since the base material class only has a shader RID reference, not a Shader resource reference.

**Comment by tetrapod00** [comment]
Maybe hold off on merging. I want to see if being exposed to scripting causes problems, and if there's a good way to avoid exposing at least the new implementation to scripting. Un-exposing the Material implementation probably breaks compat even if it's correct to do.

File: scene/resources/shader.cpp
Comment (tetrapod00) [reply scene/resources/shader.cpp:261]: Ah, looks like @clayjohn you voiced concerns in the original PR https://github.com/godotengine/godot/pull/44950. This is a niche enough feature that I'm not sure if it's worth putting a lot of time into. But I want to at least take a cursory look at possible improvements before merging this as-is.

Comment (tetrapod00) [reply scene/resources/shader.cpp:261]: Okay, this isn't flawless proof, but I [searched across all of github](https://github.com/search?q=inspect_native_shader_code&type=code) for usage of "inspect_native_shader_code", and it is 99% either forks of godot repos, or godot bindings in various languages. There is one instance of user code, in a [fuzzer for godot](https://github.com/qarmin/Qarminer/blob/cebb2e69733fcada7324f58232912a4383f40340/Autoloads/BasicData.gd#L176). I could not find any instances where a plugin with public source on github makes use of `Material.inspect_native_shader_code()` being exposed to scripting.

I don't know how high the bar is for unexposing a function to scripting is in a minor version.

Comment (tetrapod00) [reply scene/resources/shader.cpp:261]: I looked into it more, and the method needs to be added to the class DB (and exposed to scripting) in order for it to show up in the inspector of the resource. If we were *only* adding it to the shader editor, there would be no need to expose it.

See the inspector dock implementation: https://github.com/godotengine/godot/blob/842f98239713fd10cfd648cd6aa3781895f289eb/editor/inspector_dock.cpp#L566-L583

**Comment by tetrapod00** [comment]
Made the small style changes.

I also looked into whether the method needs to be exposed to the class DB and scripting, and I believe it does. The resource inspector dock uses the class DB to add methods to the `Manage Object Properties` menu. If we were *only* adding this method to the shader editor, there would be no need to expose it to scripting.

See the inspector dock implementation: https://github.com/godotengine/godot/blob/842f98239713fd10cfd648cd6aa3781895f289eb/editor/inspector_dock.cpp#L566-L583

**Comment by clayjohn** [comment]
@tetrapod00 I'm a little lost in the last few comments. Is this PR ready for merging now?

**Comment by tetrapod00** [comment]
Yes, the PR is ready to merge.

The function `Shader.inspect_native_shader_code()` is exposed to scripting, which seemed unnecessary. I looked to see if there was a way to add this action to the `Shader` resource's inspector, without exposing it to scripting, but could not find one. The rest is all details.

**Comment by Repiteo** [comment]
Thanks!

**Pull request merged**
